src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,432 @@
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.ProcessedVariables.Add(variableValue);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(variableName, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static bool ProcessAssignment(AssignmentExpressionSyntax assignmentExpression, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var evaluationResult = PreEvaluate(assignmentExpression, contextInformation, null, displayDiagnostics);
+        var variableName = (assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax).Identifier.ValueText;
+
+        if (evaluationResult.CanEvaluate)
+        {
+            var variableValue = evaluationResult.SyntaxNodeToVariableValueMap.Values.First();
+            var generateNewContext = IsCompoundExpression(assignmentExpression) || evaluationResult.SyntaxNodeToVariableValueMap.Count() > 1;
+
+            if (generateNewContext)
+            {
+                var expressionAnalysisContext = GenerateNewExpressionAnalysisContext(assignmentExpression, evaluationResult.SyntaxNodeToVariableValueMap);
+                variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+            }
+
+            dirtyVariables.Add(variableName);
+            contextInformation.CurrentVariablesValues[variableName] = variableValue;
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.ProcessedVariables.Add(variableValue);
+            }
+
+            if (assignmentExpression.Parent is ConditionalExpressionSyntax)
+            {
+                RemoveVariablesFromContext(variableName, contextInformation);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(variableName, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static void RemoveVariablesFromContext(IEnumerable<string> variablesToRemove, ContextInformation contextInformation)
+    {
+        foreach (var variable in variablesToRemove)
+        {
+            RemoveVariablesFromContext(variable, contextInformation);
+        }
+    }
+
+    private static void RemoveVariablesFromContext(string variableName, ContextInformation contextInformation)
+    {
+        var variableValues = contextInformation.CurrentVariablesValues;
+
+        if (variableValues.TryGetValueSafe(variableName, out var previousValue))
+        {
+            var originalExpression = previousValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
```

Comment:
Check if we need to add the var to ProcessedVariables

Reply:
Done